### PR TITLE
chore(nix): Fix dependency caching

### LINKED ISSUE #1303: CI should be caching cargo dependencies
### Problem

It seems that all cargo dependencies are being rebuilt on every CI run. These should be cached and rehydrated to significantly bring down the CI run times.

### Proposed solution

In talking with @kobyhallx, he suggested that maybe this is due to not using "clean sources" within Nix. I believe he has experience with this and will dive in.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by phated**
# Description

## Problem\*

Resolves #1303 

## Summary\*

This updates the flake to enable better caching. When building the dependencies, we were injecting the `GIT_COMMIT` and `GIT_DIRTY` variables, which made the dependencies rebuild from scratch on each commit. It also reworks the flake to have better delineation between native backend and wamer backend build targets (this will avoid needing the bb.wasm file during checks).

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -39,7 +39,7 @@ jobs:
         uses: actions/cache@v3
         with:
           path: /tmp/nix-cache
-          key: ${{ runner.os }}-flake-${{ hashFiles('*.lock') }}
+          key: ${{ runner.os }}-flake-${{ hashFiles('*.lock') }}-temp
```
Comment (phated) [inline .github/workflows/test.yml:42]:
```suggestion
          key: ${{ runner.os }}-flake-${{ hashFiles('*.lock') }}
```
Just needed to work with my own cache to make sure this PR is working right. Need to clear all the caches right before this is queued for merge.

**Review by kevaundray — APPROVED** [review]

**Comment by phated** [comment]
I should have switched this to a draft, as I'm still trying to look at the cache results

**Comment by kevaundray** [comment]
> I should have switched this to a draft, as I'm still trying to look at the cache results

You mean the performance difference?

**Comment by phated** [comment]
> You mean the performance difference?

Making sure I'm seeing the same results in CI as I saw locally. There's a subtle difference in how the nix store is restored in CI so I wanted to make sure this is working.

**Comment by phated** [comment]
Looks like it is working the same. Running `nix flake check` on ubuntu finishes in 10 minutes with 8 minutes 36 seconds spent in the tests.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**